improver/wxcode/weather_symbols.py

```diff
@@ -579,6 +492,165 @@ def create_symbol_cube(cubes):
+            )
+
+    def evaluate_extract_expression(self, cubes, expression):
+        """Evaluate a single condition.
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing the diagnostics required for the
+                weather symbols decision tree, these at co-incident times.
+            expression (iris.Constraint or list):
+                Defined recursively:
+                A list consisting of an iris.Constraint or a list of 
+                iris.Constraint, strings (representing operators) and floats 
+                is a valid expression.
+                A list consisting of valid expressions, strings (representing 
+                operators) and floats is a valid expression.
+        
+        Returns:
+            numpy.array:
+                An array or masked array of booleans
+        """
+        operator_map = {
+            "+": operator.add,
+            "-": operator.sub,
+            "*": operator.mul,
+            "/": operator.truediv,
+        }
+        if isinstance(expression, iris.Constraint):
+            return cubes.extract(expression)[0].data
+        else:
+            curr_expression = copy.deepcopy(expression)
+            # evaluate sub-expressions first
+            for idx, item in enumerate(expression):
+                if isinstance(item, list):
+                    curr_expression = (
+                        curr_expression[:idx]
+                        + [self.evaluate_extract_expression(cubes, item)]
+                        + curr_expression[idx + 1 :]
+                    )
+            # evaluate operators in order of precedence
+            for op_str in ["/", "*", "+", "-"]:
+                while len(curr_expression) > 1:
+                    for idx, item in enumerate(curr_expression):
+                        if isinstance(item, str) and (item == op_str):
+                            left_arg = curr_expression[idx - 1]
+                            right_arg = curr_expression[idx + 1]
+                            if isinstance(left_arg, iris.Constraint):
+                                left_eval = cubes.extract(left_arg)[0].data
+                            else:
+                                left_eval = left_arg
+                            if isinstance(right_arg, iris.Constraint):
+                                right_eval = cubes.extract(right_arg)[0].data
+                            else:
+                                right_eval = right_arg
+                            op = operator_map[op_str]
+                            res = op(left_eval, right_eval)
+                            curr_expression = (
+                                curr_expression[: idx - 1]
+                                + [res]
+                                + curr_expression[idx + 2 :]
+                            )
+                            break
+                    else:
+                        break
```

Comment:
Can't quite see what this `else` matches up with - in what circumstances would this be accessed?

Reply:
This matches with the `for` above it, it's a `for-else` block (https://docs.python.org/3/reference/compound_stmts.html#for)

The `else` part executes only if all iterations of the `for` loop complete normally. If the loop exits early by a `break` statement then the `else` condition does not execute.

In our case, the `for` loop iterates over the items in the expression searching for an operator `op_str`. If it is found, we evaluate the part involving the operator, replace `curr_expression` with the new (shorter) expression, and break out of the `for` loop. Provided the length of the expression is still > 1, the `while` condition will cause the `for` loop to start again, looking for further instances of `op_str`.

But if the `for` loop completes without finding `op_str` (i.e. the loop does not break), then we also want to break out of the `while` loop and change `op_str` to a new operator to search. This is what the `else` achieves.

Reply:
Ah, thanks for the explanation.  It's not something I'd come across before but that makes sense.